File: Examples/CantileverWetEtching/CantileverWetEtching.cpp
#include <psDomain.hpp>
#include <psGDSReader.hpp>
#include <psPlanarize.hpp>
#include <psProcess.hpp>

#include <WetEtching.hpp>

int main(int argc, char **argv) {
  using NumericType = double;
  constexpr int D = 3;

  // set number threads to be used
  omp_set_num_threads(12);

  std::string maskFileName = "cantilever_mask.gds";

  const int minutes = 120 / 5;   // total etch time (2 hours)
  const NumericType x_add = 50.; // add space to domain boundary
  const NumericType y_add = 50.;
  const NumericType gridDelta = 5.; // um

  /* -------- read GDS mask file --------- */
  typename lsDomain<NumericType, D>::BoundaryType boundaryCons[D];
  for (int i = 0; i < D - 1; i++)
    boundaryCons[i] = lsDomain<NumericType, D>::BoundaryType::
        REFLECTIVE_BOUNDARY; // boundary conditions in x and y direction
  boundaryCons[D - 1] = lsDomain<NumericType, D>::BoundaryType::
      INFINITE_BOUNDARY; // open boundary in z direction
  auto gds_mask = psSmartPointer<psGDSGeometry<NumericType, D>>::New(gridDelta);
  gds_mask->setBoundaryConditions(boundaryCons);
  gds_mask->setBoundaryPadding(x_add, y_add);
  psGDSReader<NumericType, D>(gds_mask, maskFileName).apply(); // read GDS file

  auto mask = gds_mask->layerToLevelSet(
      1 /*layer in GDS file*/, 0 /*base z position*/,
      4 * gridDelta /*mask height*/, true /*invert mask*/);

  /* -------- create plane substrate --------- */
  NumericType origin[D] = {0., 0., 0.};   // surface origin
  NumericType normal[D] = {0., 0., 1.};   // surface normal
  double *bounds = gds_mask->getBounds(); // extent of GDS mask
  auto plane = psSmartPointer<lsDomain<NumericType, D>>::New(
      bounds, boundaryCons, gridDelta);
  lsMakeGeometry<NumericType, D>(
      plane, psSmartPointer<lsPlane<NumericType, D>>::New(origin, normal))
      .apply();

  /* -------- set up geometry --------- */
  auto geometry = psSmartPointer<psDomain<NumericType, D>>::New();
  geometry->insertNextLevelSet(mask);
  geometry->insertNextLevelSet(plane);
  geometry->printSurface("InitialGeometry.vtp");

  /* -------- wet etch process --------- */
  WetEtching<NumericType, D> wetEtchModel;

  psProcess<NumericType, D> process;
  process.setDomain(geometry);
  process.setProcessModel(wetEtchModel.getProcessModel());
  process.setProcessDuration(5. * 60.); // 5 minutes of etching
  process.setPrintIntermediate(false);
  process.setIntegrationScheme(
      lsIntegrationSchemeEnum::STENCIL_LOCAL_LAX_FRIEDRICHS_1ST_ORDER);

  for (int n = 0; n < minutes; n++) {
    process.apply(); // run process
    geometry->printSurface("WetEtchingSurface_" + std::to_string(n) + ".vtp");
  }

  geometry->printSurface("FinalGeometry.vtp");
}

File: examples/cantileverWetEtching/cantileverWetEtching.py
# This example only works in 3D mode
import viennaps.d3 as psd
from viennaps import BoundaryType, Material, AdvectionParameters, DiscretizationScheme

maskFileName = "cantilever_mask.gds"

# crystal surface direction
direction100 = [0.707106781187, 0.707106781187, 0.0]
direction010 = [-0.707106781187, 0.707106781187, 0.0]
# etch rates for crystal directions in um / s
# 30 % KOH at 70°C
# https://doi.org/10.1016/S0924-4247(97)01658-0
r100 = 0.797 / 60.0
r110 = 1.455 / 60.0
r111 = 0.005 / 60.0
r311 = 1.436 / 60.0
minutes = int(120 / 5)  # total etch time (2 hours)

x_add = 50.0  # add space to domain boundary
y_add = 50.0
gridDelta = 5.0  # um

# read GDS mask file
boundaryConditions = [
    BoundaryType.REFLECTIVE_BOUNDARY,
    BoundaryType.REFLECTIVE_BOUNDARY,
    BoundaryType.INFINITE_BOUNDARY,
]

gds_mask = psd.GDSGeometry(gridDelta)
gds_mask.setBoundaryConditions(boundaryConditions)
gds_mask.setBoundaryPadding(x_add, y_add)
psd.GDSReader(gds_mask, maskFileName).apply()

# convert GDS geometry to level set
mask = gds_mask.layerToLevelSet(1, 0.0, 4 * gridDelta, True, False)

# set up domain
geometry = psd.Domain()
geometry.insertNextLevelSetAsMaterial(mask, Material.Mask)

# create plane substrate under mask
psd.MakePlane(geometry, 0.0, Material.Si, True).apply()

geometry.saveSurfaceMesh("initialGeometry.vtp", True)

# wet etch process
model = psd.WetEtching(
    direction100=direction100,
    direction010=direction010,
    rate100=r100,
    rate110=r110,
    rate111=r111,
    rate311=r311,
    materialRates=[(Material.Si, -1.0)],
)

advectionParams = AdvectionParameters()
advectionParams.discretizationScheme = (
    DiscretizationScheme.STENCIL_LOCAL_LAX_FRIEDRICHS_1ST_ORDER
)

process = psd.Process()
process.setDomain(geometry)
process.setProcessModel(model)
process.setProcessDuration(5.0 * 60.0)  # 5 minutes of etching
process.setParameters(advectionParams)

for n in range(minutes):
    # run process
    process.apply()
    geometry.saveSurfaceMesh("wetEtchingSurface_" + str(n) + ".vtp", True)

geometry.saveSurfaceMesh("finalGeometry.vtp", True)
